src/meta/src/hummock/manager/mod.rs

```diff
@@ -196,6 +200,78 @@ where
+        let join_handle = tokio::spawn(async move {
+            let mut min_interval = tokio::time::interval(std::time::Duration::from_millis(1000));
+            loop {
+                tokio::select! {
+                    // Wait for interval
+                    _ = min_interval.tick() => {
+                    },
+                    // Shutdown
+                    _ = &mut shutdown_rx => {
+                        tracing::info!("Compaction heartbeat checker is stopped");
+                        return;
+                    }
+                }
+                for (context_id, mut task) in compactor_manager.get_timed_out_tasks() {
+                    if let Some(compactor) = compactor_manager.get_compactor(context_id) {
+                        let _ = compactor.cancel_task(task.task_id).await;
+                    }
+                    // Since we cannot rely on the compactor to return, we will try to update the
+                    // compact task status ourselves which is an idempotent,
+                    // serializable operation.
+
+                    // Change task status to failed
+                    task.task_status = false;
+
+                    // TODO: this needs to be reliable, in other words, we cannot tolerate errors.
+                    // So, we retry and also have an idempotent fail-all which
+                    // we also retry (manually trigger `try_send_compaction_request`).
+                    // If all else fails, we report that we may never recover from this failure and
+                    // may fail to balance the LSM tree.
+                    let mut success = false;
+                    let mut last_err = None;
+                    for attempt_id in 0..REPORT_COMPACT_TASK_RETRIES {
+                        if let Err(e) = hummock_manager.report_compact_task(context_id, &task).await
+                        {
+                            tracing::error!("Failed to cancel compaction task due to missing heartbeat (attempt: {attempt_id}). \
+                                {context_id}, task_id: {}, ERR: {e:?}", task.task_id);
+                            last_err = Some(e);
+                        } else {
+                            success = true;
+                            break;
+                        }
+                    }
+                    if !success {
+                        tracing::error!("Failed to cancel compaction task due to missing heartbeat after {REPORT_COMPACT_TASK_RETRIES} retries. \
+                            THIS MAY RESULT IN A PERMANENTLY IMBALANCED LSM TREE. hummock context: {context_id}, task_id: {}, ERR: {:?}", task.task_id, last_err);
+                        let mut reschedule_success = false;
+                        for _ in 0..REPORT_COMPACT_TASK_RETRIES {
+                            if hummock_manager
+                                .try_send_compaction_request(task.compaction_group_id)
```

Comment:
Reschedule is optional  because eventually another write will trigger it.

Reply:
Well, I think the point is if there are no new writes. But anyway, perhaps it is good enough to retry reporting task status indefinitely and not have a restart fallback as it's quite unlikely for two errors to cooincide (can't report, and no new writes).